template/{{.input_root_dir}}/README.md.tmpl

```diff
@@ -160,17 +160,28 @@ or pipeline assets (e.g. use a larger instance type for model training) via pull
 | Data Scientist                | Update production ML code (e.g. model training logic) for an existing project | [ML pull request guide](docs/ml-pull-request.md)                                                                                                                    |
 | Data Scientist                | Modify production model ML assets, e.g. model training or inference jobs  | [ML asset config guide]({{template `project_name_alphanumeric_underscore` .}}/assets/README.md)                                                     |
 
-## Monorepo
+## Setting up CI/CD for multiple projects
 
-It's possible to use the repo as a monorepo that contains multiple projects. All projects share the same workspaces and service principals.
+It's possible to use the repo as a monorepo that contains multiple projects if all projects share the same workspaces and service principals.
 
 For example, assuming there's existing repo with root directory name `monorepo_root_dir` and project name `project1`
-1. Create another project from `databricks bundle init` with project name `project2` and root directory name `project2`.
-2. Copy the internal directory `project2/project2` to root directory of existing repo `monorepo_root_dir/project2`.
+1. Create another project from `databricks bundle init` with project name `project2` and root directory name `monorepo_root_dir`, while specifying 'no' for the 'Setup CI/CD' prompt.
 {{ if or (eq .input_cicd_platform `github_actions`) (eq .input_cicd_platform `github_actions_for_github_enterprise_servers`) -}}
-3. Copy yaml files from `project2/.github/workflows/` to `monorepo_root_dir/.github/workflows/` and make sure there's no name conflicts.
-{{ end -}}
-{{- if (eq .input_cicd_platform `azure_devops`) }}
-3. Copy yaml files from `project2/.azure/devops-pipelines/` to `monorepo_root_dir/.azure/devops-pipelines/` and make sure there's no name conflicts.
+2. Duplicate the YAML files in `monorepo_root_dir/.github/workflows/` that start with `project1-`, and be sure to change `project1` to `project2` for:
+    - File names
+    - The `'working-directory'` parameter in each file
+    - The `'pull_request'` `'paths'` parameter in the CI and tests workflow files
+{{- else if (eq .input_cicd_platform `azure_devops`) }}
+2. Duplicate the YAML files in `monorepo_root_dir/.azure/devops-pipelines/`, and be sure to change `project1` to `project2` for:
+    - File names
+    - The `'workingDirectory'` global variable in each file
+    - The `'trigger'` `'paths'` `'include'` parameter in each file
```

Comment:
It's much more work and error prone to copy all the files, rename them and update all the parameters.
It's also additional context to learn to check if the configs for the new and existing projects are the same or not and then generate the project again.
Can we always generate the CICD files with correct parameters and let users do a copy and paste?

Reply:
Yeah two reasons for this:
1. Monorepo generation was failing in the same directory before and we have that open as a GitHub issue.
2. Data scientists often don't know CI/CD parameters anyway and now the UJ for them to get started is simpler.

The copy paste approach is the one we have currently and we've seen feedback that it's confusing. Long-term we were discussing separating the architecture into two DABs anyway where the MLEs set up the infra and the data scientists can just plug and play their own project with different templates (could also be multiple DABs tbd), and then plug into the MLE infrastructure when it's ready to productionize.

Reply:
was going to make the same point about automating this for the user if they create a second project in the same repo. If it's possible for us to add the new project2 folder to the repo, could we not also add the 2 CI/CD yaml files under `devops-pipelines`, parameterized on that new project?

Reply:
if we necessarily have to have the user duplicate the YAML files, I would also call out that they need to follow the same setup steps [here](https://github.com/databricks/mlops-stacks/blob/main/template/%7B%7B.input_root_dir%7D%7D/docs/mlops-setup.md.tmpl) to create new CI/CD pipelines for the new project

Reply:
we would also want to call out in the MLOps docs [here](https://github.com/databricks/mlops-stacks/blob/main/template/%7B%7B.input_root_dir%7D%7D/docs/mlops-setup.md.tmpl#L192) that in the case of multiple projects, the build validation polices should be defined with a path filter to the respective project folders under the mono repo

Reply:
Yeah we can call out that they need to follow the setup again for ADO and we can definitely add that build validation policy section to the docs

Reply:
re the call out on path filters for build validation branch policies, let's replace [this line](https://github.com/databricks/mlops-stacks/blob/main/template/%7B%7B.input_root_dir%7D%7D/docs/mlops-setup.md.tmpl#L192) with the following:

```
1. Define two [build validation branch policies](https://learn.microsoft.com/en-us/azure/devops/repos/git/branch-policies?view=azure-devops&tabs=browser#build-validation) for the `{{template `default_branch` .}}` branch using the two Azure build pipelines created in step 1. This is required so that any PR changes to the `{{template `default_branch` .}}` must build successfully before PRs can complete. In the case of a monorepo, where there are multiple projects under a single repository, set a [path filter](https://learn.microsoft.com/en-us/azure/devops/repos/git/branch-policies?view=azure-devops&tabs=browser#path-filters) on the build validation policies, such that devops pipelines are only triggered when there are changes to the respective projects (e.g. the path filter would be `/project1/*` to trigger a devops pipeline when changes are made to _only_ files under the `project1` folder)
```

Reply:
When we generate a monorepo in the first place, can we make the generated CI/CD workflows as generic as possible so they don't even need to be copy pasted later?? Or maybe the only update the DS needs to do is to just add paths for project2 to an existing CI/CD setup? (Looking at the generated files, I guess this is not feasible, but please correct me if I'm wrong).

Reply:
> but the advantage here is that it addresses the feedback that DS usually don't have CI/CD information and keeps things simpler for them
 
Are we addressing the case where users are adding a new project to an existing repo? In that case CI/CD has already been setup for the previous projects though. I wonder if we can automate the setup in that case (and add a bunch of TODOs where it needs updating) for them.

Or the case where there is no CI/CD at all (in that case how would it be added later?)


Should the options be like: cicd_setup -> use existing/none/generate CICD workflows from scratch.

Reply:
So they can't just use existing natively since the CICD workflows are project-specific, but only based on the working directory. That's pretty much the only difference. 

We could generate the generic workflows but then we'd need to ask for `cicd_platform` at least as a parameter, and we'd have to assume `main` branch name, `release` branch name, `staging_workspace_host`, and `production_workspace_host`, marking as TODO for later on. Is that a better experience?

Reply:
Interesting. We can completely decouple CICD from the project. The current problem is that a workflow is project-specific and needs manual modification. To resolve that, we can make the github/ado workflow file auto-generated from a more generic "workflow-generation workflow".

---

### 1. Create a data asset bundle template for generating CICD yml files for a specific project.
The template input variables are the following
- project_name
- cicd_platform
- main_branch
- release_branch
- staging_workspace_host
- production_workspace_host
- is_fs_enabled
- is_recipes_enabled

The output files are the CICD yaml files with project name in the prefix and proper files skipped based on input.

### 2. Add a yaml file in each project dir so that the "workflow-generation workflow" can load project_name, is_fs_enabled and is_recipes_enabled
The file can be something like: `<root_dir>/<project_name>/.proj_cicd_conf`
```
project_name : my_project
is_fs_enabled : yes
is_recipes_enabled : no
```

### 3. Add a yaml file in root dir so that the "workflow-generation workflow" can load CICD configs
The file can be something like: `<root_dir>/<project_name>/.root_cicd_conf`
```
cicd_platform : github
main_branch : main
release_branch : release
staging_workspace_host : www.staging.com
production_workspace_host : www.prod.com
```

### 4. Add the "workflow-generation workflow"
1. The workflow is triggered whenever there's a PR or commit.
2. read the `.root_cicd_conf`
3. loop through first level sub dirs and try to find the `.proj_cicd_conf`
4. For each project dir with `.proj_cicd_conf`, call bundle template to generate and workflow yml files and copy to the workflow dir.
5. If there's any update in the repo, commit and push the change.

---

Another benefit is that, now it's also possible to use a slightly different version of the CICD bundle template to add CICD to an existing project-only stacks prject.

Reply:
Yeah @sebastiaanboer-db and I were talking about this, but idk how it works if we go down the multiple templates route as of now. We'd have to figure out a few things (such as what initialization looks like, one point-of-entry switching to multiple, how this will be exposed to users via the CLI/UI, etc.) and there are a few related drawbacks to it as well.

In terms of the workflow you've proposed, we'd have to commit changes back to the repo as you said. So technically there would be a delay while the workflow-generation workflow runs, commits the changes back, then the project-specific workflow files run again (hopefully, not sure if committed workflows run retroactively on open PRs). Might be better to run this workflow only on PRs when one of the conf files is modified. But I'm not sure if this way is better, I think it does complicate things and requires us to create a separate template/repo

Reply:
To clarify, we are basically addressing the case when DS is creating new projects in existing repo here, right? I like @mingyu89 's direction - I wonder if we can do something along those lines without revamping everything (yet).
For instance, we store the cicd conf on first initialization and reuse it later to automatically generate CICD workflow configs as new projects are added? We can discuss exactly how to setup this automation (this is all part of bundle init?)

(this CICD conf can become the basis for a separate template in the future)